### PR TITLE
Move maxConcurrency to evaluator classes

### PR DESCRIPTION
**Pull request by nicolewhite**

### CODE REVIEW

File: src/testing/util.ts [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * concurrent operations for a given task.
+ *
+ * https://github.com/python/cpython/blob/5dc8c84d397110f9edfa56793ad8887b1f176d79/Lib/asyncio/locks.py#L352
+ */
+export class Semaphore {
+  private available: number;
+  private waiters: Set<Waiter>;
+
+  constructor(max: number) {
+    this.available = max;
```
Comment (adamnolte) [inline src/testing/util.ts:17]: What do you think about computing available based on max and # of waiters? That way we don't have to separately track it.

Comment (nicolewhite) [reply src/testing/util.ts:17]: we don't actually start adding waiters until available is 0, so i think it needs to be tracked separately: 

```ts
  private async acquire(): Promise<void> {
    if (this.available > 0) {
      this.available--;
      return;
    }
```

Comment (nicolewhite) [reply src/testing/util.ts:17]: btw was following this if it's helpful: https://github.com/python/cpython/blob/5dc8c84d397110f9edfa56793ad8887b1f176d79/Lib/asyncio/locks.py#L352

File: src/testing/util.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        await promise;
+      } finally {
+        this.waiters.delete(waiter);
+      }
+    } finally {
+      while (this.available > 0) {
+        if (!this.wakeUpNext()) {
+          break;
+        }
+      }
```
Comment (adamnolte) [inline src/testing/util.ts:53]: What is the situation where this is needed?

Wouldn't release always wake up the next waiter?

Comment (nicolewhite) [reply src/testing/util.ts:53]: hmm good question, i was following this but yeah i guess i'm not sure why you need this + the logic in release

https://github.com/python/cpython/blob/5dc8c84d397110f9edfa56793ad8887b1f176d79/Lib/asyncio/locks.py#L420-L425

```python
        finally:
            # New waiters may have arrived but had to wait due to FIFO.
            # Wake up as many as are allowed.
            while self._value > 0:
                if not self._wake_up_next():
                    break  # There was no-one to wake up.
```

Comment (adamnolte) [reply src/testing/util.ts:53]: The reason I ask is I was trying to think through if this is going to result in an infinite loop in some situations

Comment (adamnolte) [reply src/testing/util.ts:53]: I guess not if they are using it there

**Review by adamnolte — APPROVED** [review]

**Review by adamnolte — APPROVED** [review]

**Pull request merged**